lib/logitech_receiver/hidpp20.py

```diff
@@ -357,53 +359,276 @@ def __len__(self):
+        if persist is not None and special_keys.KEY_FLAG.persistently_divertable not in self.flags:
+            raise FeatureNotSupported(
+                'Tried to persistently divert non-persistently-divertable control ' + f'{self.key} on device {self._device}.'
+            )
+        if rawXY is not None and not self.rawXY_reportable:
+            raise FeatureNotSupported(
+                f'Tried to request raw XY reports from control {self.key} with ' +
+                f'no raw XY capability on device {self._device}.'
+            )
+        if remap != 0 and remap not in self.remappable_to:
+            raise FeatureNotSupported(
+                f'Tried to remap control {self.key} to a control ID {remap} which it is not remappable to ' +
+                f'on device {self._device}.'
+            )
+
+        mkbit = lambda v: 1 if v else 0
+        isset = lambda v: mkbit(v is not None)
+
+        pkt = tuple(
+            _pack(
+                '!HBH',
+                self._cid,
+                (isset(rawXY) << 5)
+                | (mkbit(rawXY) << 4)
+                | (isset(persist) << 3)
+                | (mkbit(persist) << 2)
+                | (isset(divert) << 1)
+                | mkbit(divert),
+                remap,
+            )
+        )
+        ret = feature_request(self._device, FEATURE.REPROG_CONTROLS_V4, 0x30, *pkt)
+        if _unpack('!BBBBB', ret[:5]) != pkt and _log.isEnabledFor(_WARNING):
+            _log.warn(
+                f'REPROG_CONTROLS_v4 endpoint setCidReporting on device {self._device} should echo request ' +
+                "packet, but didn't."
+            )
+
+        # update knowledge of mapping
+        self._getCidReporting()
+
+
 class KeysArray(object):
     """A sequence of key mappings supported by a HID++ 2.0 device."""
-    __slots__ = ('device', 'keys', 'keyversion')
+
+    __slots__ = ('device', 'keys', 'keyversion', 'cid_to_tid', 'group_cids')
 
     def __init__(self, device, count):
         assert device is not None
         self.device = device
-        self.keyversion = 0
+        if FEATURE.REPROG_CONTROLS in self.device.features:
+            self.keyversion = 1
+        elif FEATURE.REPROG_CONTROLS_V4 in self.device.features:
+            self.keyversion = 4
+        else:
+            if _log.isEnabledFor(_ERROR):
+                _log.error(f'Trying to read keys on device {device} which has no REPROG_CONTROLS(_VX) support.')
+            self.keyversion = None
         self.keys = [None] * count
+        # The mapping from Control IDs to their native Task IDs.
+        # For example, Control "Left Button" is mapped to Task "Left Click".
+        # When remapping controls, we point the control we want to remap
+        # at a target Control ID rather than a target Task ID. This has the
+        # effect of performing the native task of the target control,
+        # even if the target itself is also remapped. So remapping
+        # is not recursive.
```

Comment:
Use long strings for comments that span multiple lines.

```
'''

'''
```

Reply:
I think that v4 is downward-compatible with previous versions - earlier versions will have 0 for the extra flags so any software will know that these parts of the feature are not available.

Reply:
(Resolved long string comments, but keeping open for v4 discussion.)

To be clear, we are talking about `REPROG_CONTROLS_V4` version 1 vs version 4 (yay naming), right? My take on this is:
- For all `get` functions, we can be backwards compatible by reading `unused` bits which should hopefully be 0, and if the device doesn't send enough bytes because a later spec version added another byte to the response format, we just pretend said byte was NUL.
- Notification events aren't a problem, older spec implementations just won't send them.
- But `setCidReporting` seems problematic. Version 4 seems to have added a fifth byte for `analyticsKeyEvt`. What happens if a device implementing earlier versions gets this byte? Also what happens if we set bits marked `unused` in version 1, because they have some function in version 4?

Reply:
Yes, newer versions are backward-compatible but they can introduce new fields, if we present values from newer versions to the users, they will have a bad experience on older versions.

Acording to your description, I think we can have a hybrid implementation that is aware from v4 for everything other than `setCidReporting`.

Reply:
@FFY00 There are two "versions" here - the feature (REPROG_CONTROLS vs REPROG_CONTROLS_V4) and the feature version (versions 0 through 4 of REPROG_CONTROLS_V4).  Everything in Solaar understands the feature differences  - the data structures keep track of the feature (class ReprogrammableKey vs ReprogrammableKeyV4, setting template checks for REPROG_CONTROLS_V4, etc).  The feature version is also taken care of because there are flags in the Cid structure that tell what can be done with the control and devices that don't support higher feature versions will have these flags false.